File: lib/services/lead-service.ts
/**
 * LeadService - Serviço unificado para criação e busca de leads
 * Suporta deduplicação cross-source (recebearquivos + socialwiseflow)
 */

import { getPrismaInstance } from "@/lib/connections";
import { LeadSource, type Lead, type Chat } from "@prisma/client";

export interface FindOrCreateLeadOptions {
	// Identificadores primários (pelo menos um obrigatório)
	phoneNumber?: string;
	chatwitContactId?: string;

	// Contexto
	chatwitAccountId: string;
	inboxId?: string;

	// Dados de enriquecimento
	name?: string;
	email?: string;
	avatarUrl?: string;
}

export interface LeadWithChat {
	lead: Lead;
	chat: Chat;
	created: boolean;
}

export class LeadService {
	private get prisma() {
		return getPrismaInstance();
	}

	/**
	 * Busca ou cria um lead por telefone OU Chatwit contact ID
	 * Suporta ambas as fontes: recebearquivos (OAB) e socialwiseflow (WhatsApp)
	 */
	async findOrCreateLead(options: FindOrCreateLeadOptions): Promise<LeadWithChat> {
		const accountId = `CHATWIT_${options.chatwitAccountId}`;

		// Estratégia 1: Tentar encontrar por telefone (cross-source)
		if (options.phoneNumber) {
			const normalizedPhone = this.normalizePhone(options.phoneNumber);
			const existingByPhone = await this.prisma.lead.findFirst({
				where: {
					phone: normalizedPhone,
					accountId,
				},
			});

			if (existingByPhone) {
				const chat = await this.findOrCreateChat(existingByPhone.id, accountId);
				return { lead: existingByPhone, chat, created: false };
			}
		}

		// Estratégia 2: Tentar encontrar por Chatwit contact ID (sourceIdentifier)
		if (options.chatwitContactId) {
			const existingByContactId = await this.prisma.lead.findFirst({
				where: {
					sourceIdentifier: options.chatwitContactId,
					accountId,
				},
			});

			if (existingByContactId) {
				const chat = await this.findOrCreateChat(existingByContactId.id, accountId);
				return { lead: existingByContactId, chat, created: false };
			}
		}

		// 🔒 VALIDAÇÃO: Verificar se Account existe antes de criar Lead (prevenir FK error)
		const accountExists = await this.prisma.account.findUnique({
			where: { id: accountId },
			select: { id: true },
		});

		if (!accountExists) {
			console.error("🚨 ERRO: Tentativa de criar Lead com Account inexistente", {
				accountId,
				chatwitAccountId: options.chatwitAccountId,
				solucao: [
					"1. Criar Account com id=%s no banco de dados",
					"2. Ou vincular inbox a uma Account existente",
					"3. Verificar tabela UsuarioChatwit para chatwitAccountId correto"
				],
				phoneNumber: options.phoneNumber,
				chatwitContactId: options.chatwitContactId,
			});
			throw new Error(
				`Account ${accountId} não existe. Impossível criar Lead. Configure a Account primeiro ou vincule a inbox a uma Account válida.`
			);
		}

		// Criar novo lead com source apropriado
		const source = options.inboxId ? LeadSource.WHATSAPP_SOCIAL_FLOW : LeadSource.CHATWIT_OAB;
		const sourceIdentifier = options.chatwitContactId || options.phoneNumber || `auto_${Date.now()}`;

		const newLead = await this.prisma.lead.create({
			data: {
				name: options.name || "Lead sem nome",
				phone: options.phoneNumber ? this.normalizePhone(options.phoneNumber) : null,
				email: options.email,
				avatarUrl: options.avatarUrl,
				source,
				sourceIdentifier: String(sourceIdentifier),
				accountId,
				tags: [],
			},
		});

		const chat = await this.findOrCreateChat(newLead.id, accountId);
		return { lead: newLead, chat, created: true };
	}

	/**
	 * Busca lead existente por telefone
	 */
	async findLeadByPhone(phoneNumber: string, accountId: string): Promise<Lead | null> {
		const normalizedPhone = this.normalizePhone(phoneNumber);
		return this.prisma.lead.findFirst({
			where: {
				phone: normalizedPhone,
				accountId,
			},
		});
	}

	/**
	 * Busca lead existente por sourceIdentifier
	 */
	async findLeadBySourceIdentifier(sourceIdentifier: string, accountId: string): Promise<Lead | null> {
		return this.prisma.lead.findFirst({
			where: {
				sourceIdentifier,
				accountId,
			},
		});
	}

	/**
	 * Atualiza o updatedAt do lead para refletir atividade recente
	 */
	async touchLead(leadId: string): Promise<void> {
		await this.prisma.lead.update({
			where: { id: leadId },
			data: { updatedAt: new Date() },
		});
	}

	/**
	 * Busca ou cria Chat para um lead
	 */
	private async findOrCreateChat(leadId: string, accountId: string): Promise<Chat> {
		const existing = await this.prisma.chat.findUnique({
			where: { leadId_accountId: { leadId, accountId } },
		});

		if (existing) return existing;

		return this.prisma.chat.create({
			data: { leadId, accountId },
		});
	}

	/**
	 * Normaliza número de telefone para formato consistente
	 */
	private normalizePhone(phone: string): string {
		// Remove tudo exceto números e '+'
		return phone.replace(/[^\d+]/g, "");
	}
}

// Singleton para uso em toda a aplicação
export const leadService = new LeadService();

File: platform-backend/domains/socialwise/services/leads/lead_service.py
"""LeadService — unified lead creation and lookup with cross-source dedup.

Port of: lib/services/lead-service.ts
"""

from __future__ import annotations

import re
from dataclasses import dataclass
from typing import Any

from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession

from domains.socialwise.db.models.account import Account
from domains.socialwise.db.models.chat import Chat
from domains.socialwise.db.models.lead import Lead, LeadSource
from domains.socialwise.db.base import generate_cuid
from platform_core.logging.config import get_logger

logger = get_logger(__name__)

_PHONE_RE = re.compile(r"[^\d+]")


@dataclass(slots=True)
class FindOrCreateLeadOptions:
    chatwit_account_id: str
    phone_number: str | None = None
    chatwit_contact_id: str | None = None
    inbox_id: str | None = None
    name: str | None = None
    email: str | None = None
    avatar_url: str | None = None


@dataclass(slots=True)
class LeadWithChat:
    lead: Lead
    chat: Chat
    created: bool


class LeadService:
    """Stateless service — receives session per call."""

    @staticmethod
    def normalize_phone(phone: str) -> str:
        return _PHONE_RE.sub("", phone)

    async def find_or_create_lead(
        self,
        session: AsyncSession,
        options: FindOrCreateLeadOptions,
    ) -> LeadWithChat:
        account_id = f"CHATWIT_{options.chatwit_account_id}"

        # Strategy 1: find by phone (cross-source)
        if options.phone_number:
            normalized = self.normalize_phone(options.phone_number)
            result = await session.execute(
                select(Lead).where(Lead.phone == normalized, Lead.account_id == account_id).limit(1)
            )
            existing = result.scalar_one_or_none()
            if existing:
                chat = await self._find_or_create_chat(session, existing.id, account_id)
                return LeadWithChat(lead=existing, chat=chat, created=False)

        # Strategy 2: find by Chatwit contact ID
        if options.chatwit_contact_id:
            result = await session.execute(
                select(Lead).where(
                    Lead.source_identifier == options.chatwit_contact_id,
                    Lead.account_id == account_id,
                ).limit(1)
            )
            existing = result.scalar_one_or_none()
            if existing:
                chat = await self._find_or_create_chat(session, existing.id, account_id)
                return LeadWithChat(lead=existing, chat=chat, created=False)

        # Validate Account exists before FK write
        account_exists = await session.execute(
            select(Account.id).where(Account.id == account_id).limit(1)
        )
        if not account_exists.scalar_one_or_none():
            raise ValueError(
                f"Account {account_id} não existe. Impossível criar Lead. "
                "Configure a Account primeiro ou vincule a inbox a uma Account válida."
            )

        source = LeadSource.WHATSAPP_SOCIAL_FLOW if options.inbox_id else LeadSource.CHATWIT_OAB
        source_identifier = (
            options.chatwit_contact_id
            or options.phone_number
            or f"auto_{generate_cuid()}"
        )

        new_lead = Lead(
            id=generate_cuid(),
            name=options.name or "Lead sem nome",
            phone=self.normalize_phone(options.phone_number) if options.phone_number else None,
            email=options.email,
            avatar_url=options.avatar_url,
            source=source.value,
            source_identifier=str(source_identifier),
            account_id=account_id,
            tags=[],
        )
        session.add(new_lead)
        await session.flush()

        chat = await self._find_or_create_chat(session, new_lead.id, account_id)
        return LeadWithChat(lead=new_lead, chat=chat, created=True)

    async def find_lead_by_phone(
        self, session: AsyncSession, phone_number: str, account_id: str,
    ) -> Lead | None:
        normalized = self.normalize_phone(phone_number)
        result = await session.execute(
            select(Lead).where(Lead.phone == normalized, Lead.account_id == account_id).limit(1)
        )
        return result.scalar_one_or_none()

    async def find_lead_by_source_identifier(
        self, session: AsyncSession, source_identifier: str, account_id: str,
    ) -> Lead | None:
        result = await session.execute(
            select(Lead).where(
                Lead.source_identifier == source_identifier,
                Lead.account_id == account_id,
            ).limit(1)
        )
        return result.scalar_one_or_none()

    async def touch_lead(self, session: AsyncSession, lead_id: str) -> None:
        from datetime import datetime, timezone

        result = await session.execute(select(Lead).where(Lead.id == lead_id).limit(1))
        lead = result.scalar_one_or_none()
        if lead:
            lead.updated_at = datetime.now(timezone.utc)

    async def _find_or_create_chat(
        self, session: AsyncSession, lead_id: str, account_id: str,
    ) -> Chat:
        result = await session.execute(
            select(Chat).where(Chat.lead_id == lead_id, Chat.account_id == account_id).limit(1)
        )
        existing = result.scalar_one_or_none()
        if existing:
            return existing

        chat = Chat(id=generate_cuid(), lead_id=lead_id, account_id=account_id)
        session.add(chat)
        await session.flush()
        return chat


lead_service = LeadService()
